### ISSUE TITLE
Validation split can overlap training data after interleave oversampling

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#680]
## Description

With non-streaming multi-dataset training, the `over` strategy and some `probs`
configurations duplicate rows before Soup performs the automatic `val_split`. Exact copies
of the same example can therefore appear in both training and validation.

Affected revision: `ef2e02b51a1b34546823edce55715d28234805cb` (Soup 0.74.0).

Relevant code:

- [`_cycle_to` and `_combine_interleaved`](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/data/loader.py#L447-L514)
- [combination before `_finalize`](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/data/loader.py#L517-L572)
- [automatic validation slice](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/data/loader.py#L235-L259)

## Steps to reproduce

Use two distinct synthetic sources:

- A: `A0 ... A7`
- B: `B0, B1`
- `data.interleave: over`
- `data.val_split: 0.1`

Execute the real `_combine_interleaved()` result through `_finalize()`.

```text
combined rows: 16
training rows: 14
validation rows: B0, B1
validation rows also present in training: B0, B1
```

The same probe with `probs=(0.5, 0.5)` also places validation row `B0` in training.
Controls without repetition, and a control that reserves validation before oversampling,
have no overlap.

The problem can be substantially worse than this small example. With source A containing
10,000 rows, source B containing 1,000 rows, `over`, and `val_split: 0.1`, all 2,000
validation rows can also occur in the 18,000-row training partition.

The maintainer independently confirmed the overlap through the public `load_dataset(cfg.data)`
path on revision `2edc063`.

## Scope

This affects both local multi-dataset loading and the Hub interleave fallback. In
`_load_interleaved_hub_datasets()`, when only some Hub entries provide their own `validation`
split, the remaining combined data still goes through `_finalize()` after repetition and can
leak into the automatically created validation set.

There is no supported `data.val` field for supplying an independent local validation dataset;
unknown configuration keys are ignored. A Hub dataset's existing `validation` split is the only
current path that avoids this particular fallback for that dataset.

## Expected behavior

Automatic validation data must contain no logical example already present in training because
of Soup's own oversampling.

## Impact

Validation loss can be arbitrarily optimistic and can mislead checkpoint or hyperparameter
selection. On local interleaved data there is no configuration workaround other than avoiding
`over`/affected `probs` combinations or disabling the automatic validation split.

## Proposed fix and acceptance criteria

Reserve validation from each source's unique rows before any cycling or probabilistic repetition,
then apply repetition only to the training partitions. Preserve a Hub-provided validation split
without modification.

- [ ] `set(train_ids).isdisjoint(validation_ids)` holds for `over` and affected `probs` cases;
      row counts alone are not sufficient.
- [ ] Both source orders are covered.
- [ ] Local interleave and the partial-Hub-validation fallback are both covered.
- [ ] Existing Hub `validation` splits remain unchanged and excluded from training.
- [ ] `concat` and `under` behavior remains unchanged.
- [ ] A regression test fails if repetition is moved back before the split.

## Implementation status

PR [#701](https://github.com/MakazhanAlpamys/Soup/pull/701) is open and targets both the local
and Hub-fallback paths with set-intersection regression tests.

## Environment

Initially reproduced with the original pure-Python functions from the pinned checkout under
Python 3.12.2. The maintainer also confirmed it through the real public loader path. No model or
training backend is involved.

**Comment by MakazhanAlpamys** [comment]
**Confirmed — and worse than you filed it.** Reproduced through the real public `load_dataset(cfg.data)` on `2edc063`, not the private helpers:

```
Interleaved 2 datasets (strategy=over) -> 16 rows
  strategy=over,  val_split=0.1 : val ALSO in train = 2/2  -> ['B0','B1']
  strategy=probs [.5,.5]        : val ALSO in train = 1/1  -> ['B0']
  CONTROL concat: 0/1     CONTROL under: 0/1
```

Your mechanism is exactly right: `_combine_interleaved` cycles the smaller source up to `max(sizes)` via `_cycle_to`, then `_finalize` takes a **positional** tail slice — `formatted[:split_idx]` / `formatted[split_idx:]` — with no dedup and no awareness that the tail it is slicing is duplicated material.

## Two amendments, both against your own framing

**At realistic ratios the leak is total, not partial.** Your example shows 2 of 2; scaled up:

```
A=10000  B=1000  over  val_split=0.1
  train=18000  val=2000  val-rows-also-in-train = 2000/2000  (100.0%)
```

Every validation row is in the training set. Validation loss on the oversampled source is measuring memorisation, and it will look *good* — which is the worst direction for a silent defect, because a suspiciously low val loss reads as success.

**The mitigation you offer does not exist on this path.** The issue suggests an independently supplied validation set avoids it. There is no `data.val` field:

```
[f for f in DataConfig.model_fields if 'val' in f]  ->  ['val_split', 'eval_on_each_dataset']
Warning: unknown config key 'data.val' - not applied.
```

An independent val set is only reachable through a *Hub* dataset's own `validation` split. For local interleave there is no escape hatch at all — the only workaround today is not using `over` or `probs`.

That second point matters for whoever fixes this, because it rules out "document the workaround" as an answer.

## Scope note for a fixer

`_load_interleaved_hub_datasets` inherits the same defect whenever only *some* entries carry a hub validation split, since it falls back to the same `_finalize` positional slice.

The only console output on the whole path is `Interleaved 2 datasets (strategy=over) -> 16 rows`. Nothing warns.

Labelled `bug` + `help wanted`. Unclaimed — comment here naming it before opening a PR. Whoever takes it: the acceptance test should assert **zero set-intersection between train and val rows**, not merely that the counts look right; a positional-slice fix that still duplicates would pass a count assertion.

**Comment by AmirF194** [comment]
Taking this, including the hub-path scope note (I'll fix the partial-hub-val-split fallback in `_load_interleaved_hub_datasets` too, same underlying defect).

**Comment by MakazhanAlpamys** [comment]
Linking the pull request onto the issue, which I should have done when it appeared: **#701** by @AmirF194 — and you were right to take the hub-path scope note with it; both paths are fixed and pinned independently.

Measured through the real `load_dataset(cfg.data)`, `LEAK` being distinct val rows that also appear in train:

| config | main | #701 |
|---|---|---|
| `[10000, 1000]` `over` 0.1 | **1000 / 1000 (100%)** | **0 / 1100** |
| `[10000, 1000]` `probs [.5,.5]` 0.1 | **1000 / 1000 (100%)** | **0 / 1100** |
| control: `concat` / `under` | 0 | 0, counts byte-identical |

So this issue was **understated**, as I noted when I confirmed it: at realistic ratios the leak is total, not partial.

Two things this produced beyond the fix. **#702** is the streaming half, which #701 provably cannot reach — on that branch `streaming: true` + `over` + `val_split 0.2` still leaks 10 of 30 distinct val rows — and it is open and unclaimed. And a consequence of the correct fix that is now documented rather than left to be discovered: train and val come from **different mixtures** once `over`/`probs` pads. On `1000/100, over, 0.1` train is 50.0% source A while val is 90.9% source A, so someone oversampling specifically to rebalance now gets a validation set in the original skew.

**Comment by MakazhanAlpamys** [comment]
Fixed in `dfa8615` by @AmirF194 in #701.

Try:

```bash
soup train --config soup.yaml    # with data.interleave.strategy: over and data.val_split > 0
```

Validation is now carved out **per source, before padding**, on the local and hub paths alike. Measured through the real `load_dataset(cfg.data)`, LEAK being distinct val rows that also appear in train:

| config | main | fixed |
|---|---|---|
| `[10000, 1000]` `over` 0.1 | **1000 / 1000 (100%)** | **0 / 1100** |
| `[10000, 1000]` `probs [.5,.5]` 0.1 | **1000 / 1000 (100%)** | **0 / 1100** |
| control `concat` / `under` | 0 | 0, counts byte-identical |

**This issue was understated**, as I said when confirming it: at realistic ratios the leak is total, not partial.

Six mutations killed, and the local and hub halves are pinned independently — breaking only the local gate fails 4 tests while the hub test stays green.

**One thing from review worth keeping, because it is a general lesson.** The test named after this issue was *vacuous*: `assert not (train_texts & val_texts)` is trivially true against an empty val set, and its companion assertion was a **total**, which is invariant under never-splitting. It passed on a build where validation did not happen at all. One added `assert val_texts` and it now fails alone under that mutation. A disjointness assertion needs its non-empty subject asserted first, or it is not evidence.

**Two consequences of the correct fix are now documented rather than left to be discovered:** a source that `val_split` empties raises with the real cause instead of a generic "must have >= 1 row"; and train and val come from **different mixtures** once `over`/`probs` pads — on `1000/100, over, 0.1` train is 50.0% source A while val is 90.9% — so someone oversampling specifically to rebalance gets a validation set in the original skew. That is in `docs/data.md`.

**Scope: the streaming path is not fixed and cannot be by this.** On the merged code, `streaming: true` + `over` + `val_split 0.2` still leaks 10 of 30 distinct val rows. That is #702, open and unclaimed, and it needs a genuinely different mechanism.

Reported by @Amix29.
